<a href="https://colab.research.google.com/github/EvgeniyGagen/EVG/blob/main/Zadacha3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [16]:
import pandas as pd

In [17]:
df = pd.read_csv('https://gbcdn.mrgcdn.ru/uploads/asset/5087678/attachment/a3a1af2058ad08de6ab63618c1dfeb9f.csv', sep=',')
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

1.1 Создать новый признак price_per_sq_lot, который будет содержать среднюю стоимость за один кв. метр общей площади

In [18]:
df['price_per_sq_lot'] = df['sqft_lot'].mean()
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...        1180              0   
1      7242     2.0           0     0  ...        2170            400   
2     10000     1.0           0     0  ...         770              0   
3      5000     1.0           0     0  ...        1050            910   
4      8080     1.0           0     0  ...        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  price_per_sq_lot  
0        5650      15106.967566  
1        7639      15106.967566  
2        8062      15106.967566  
3        5000      15106.967566  
4        7503      15106.967566  

[5 rows x 22 columns]

1.2 Создать новый признак delta_renovated, который будет содержать разницу в годах между годом реновацией дома и годом постройки дома
Если реновации дома не было, то в новом признаке поставьте 0

In [19]:
A = []
for i, row in df.iterrows():
    if row['yr_renovated']:
        A.append(row['yr_renovated']-row['yr_built'])
    else:
        A.append(0)

df['delta_renovated'] = A
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  sqft_basement  yr_built  \
0      5650     1.0           0     0  ...              0      1955   
1      7242     2.0           0     0  ...            400      1951   
2     10000     1.0           0     0  ...              0      1933   
3      5000     1.0           0     0  ...            910      1965   
4      8080     1.0           0     0  ...              0      1987   

   yr_renovated  zipcode      lat     long  sqft_living15  sqft_lot15  \
0             0    98178  47.5112 -122.257           1340        5650   
1          1991    98125  47.7210 -122.319           1690        7639   
2             0    98028  47.7379 -122.233           2720        8062   
3             0    98136  47.5208 -122.393           1360        5000   
4             0    98074  47.6168 -122.045           1800        7503   

   price_per_sq_lot  delta_renovated  
0      15106.967566                0  
1      15106.967566               40  
2      15106.967566                0  
3      15106.967566                0  
4      15106.967566                0  

[5 rows x 23 columns]

1.3 Создайте признаки года продажи, месяца продажи

In [20]:
df['year_sale'] = df['date'].apply(lambda x: int(x[:4]))
df['month_sale'] = df['date'].apply(lambda x: int(x[4:6]))

df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  yr_renovated  zipcode      lat  \
0      5650     1.0           0     0  ...             0    98178  47.5112   
1      7242     2.0           0     0  ...          1991    98125  47.7210   
2     10000     1.0           0     0  ...             0    98028  47.7379   
3      5000     1.0           0     0  ...             0    98136  47.5208   
4      8080     1.0           0     0  ...             0    98074  47.6168   

      long  sqft_living15  sqft_lot15  price_per_sq_lot  delta_renovated  \
0 -122.257           1340        5650      15106.967566                0   
1 -122.319           1690        7639      15106.967566               40   
2 -122.233           2720        8062      15106.967566                0   
3 -122.393           1360        5000      15106.967566                0   
4 -122.045           1800        7503      15106.967566                0   

   year_sale  month_sale  
0       2014          10  
1       2014          12  
2       2015           2  
3       2014          12  
4       2015           2  

[5 rows x 25 columns]

1.4 Удалите признаки date, zipcode, lat, long

In [21]:
df.drop(columns=['date', 'zipcode', 'lat', 'long'], inplace=True)
df.head()

id     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0  7129300520  221900.0         3       1.00         1180      5650     1.0   
1  6414100192  538000.0         3       2.25         2570      7242     2.0   
2  5631500400  180000.0         2       1.00          770     10000     1.0   
3  2487200875  604000.0         4       3.00         1960      5000     1.0   
4  1954400510  510000.0         3       2.00         1680      8080     1.0   

   waterfront  view  condition  ...  sqft_above  sqft_basement  yr_built  \
0           0     0          3  ...        1180              0      1955   
1           0     0          3  ...        2170            400      1951   
2           0     0          3  ...         770              0      1933   
3           0     0          5  ...        1050            910      1965   
4           0     0          3  ...        1680              0      1987   

   yr_renovated  sqft_living15  sqft_lot15  price_per_sq_lot  delta_renovated  \
0             0           1340        5650      15106.967566                0   
1          1991           1690        7639      15106.967566               40   
2             0           2720        8062      15106.967566                0   
3             0           1360        5000      15106.967566                0   
4             0           1800        7503      15106.967566                0   

   year_sale  month_sale  
0       2014          10  
1       2014          12  
2       2015           2  
3       2014          12  
4       2015           2  

[5 rows x 21 columns]

Условие 2:

Создайте датафрейм с клиентами:

clients = pd.DataFrame({

'client_id': [1459, 4684, 3498, 3942, 4535, 2308, 2866, 2765, 1472, 4236, 2295,
939, 3840, 280, 20, 4332, 3475, 4213, 3113, 4809, 2134, 2242,
2068, 4929, 1384, 1589, 3317, 2260, 1727, 1764, 1611, 1474],

'house_id': [8965450190, 6823100225, 5104540330, 2131701075, 1522700060,
1189000207, 6821600300, 7137950720, 9510920050, 6131600255,
5428000070, 1788800910, 8100400160, 3123049142, 6306800010,
5083000375, 7920100025, 1951600150, 809001400, 339600110,
1622049154, 1099600250, 8563000110, 2768100205, 3995700435,
8861700030, 3303980210, 7731100066, 8146100580, 825069097,
3889100029, 9524100196]
})

In [22]:
clients = pd.DataFrame({

'client_id': [1459, 4684, 3498, 3942, 4535, 2308, 2866, 2765, 1472, 4236, 2295, 939, 3840, 280, 20, 4332, 3475, 4213, 3113, 4809, 2134, 2242, 2068, 4929, 1384, 1589, 3317, 2260, 1727, 1764, 1611, 1474],
 })
clients.head()

client_id    house_id
0       1459  8965450190
1       4684  6823100225
2       3498  5104540330
3       3942  2131701075
4       4535  1522700060

In [25]:
house = pd.DataFrame({'house_id': [8965450190, 6823100225, 5104540330, 2131701075, 1522700060, 1189000207, 6821600300, 7137950720, 9510920050, 6131600255, 5428000070, 1788800910, 8100400160, 3123049142, 6306800010, 5083000375, 7920100025, 1951600150, 809001400, 339600110, 1622049154, 1099600250, 8563000110, 2768100205, 3995700435, 8861700030, 3303980210, 7731100066, 8146100580, 825069097, 3889100029, 9524100196]})
house.head()

house_id
0  8965450190
1  6823100225
2  5104540330
3  2131701075
4  1522700060

2.1 Присоедините к таблице clients данные по домам через метод join

In [24]:
clients_id = clients.set_index('house_id')
df_id = df.set_index('id')

clients_id.join(df_id).reset_index().head()


index  client_id     price  bedrooms  bathrooms  sqft_living  \
0   339600110       4809  395000.0         3       2.50         1610   
1   809001400       3113  925000.0         3       1.00         1630   
2   825069097       1764  770000.0         3       2.50         2650   
3  1099600250       2242  260000.0         3       1.75         1710   
4  1189000207       2308  387000.0         2       2.50         1170   

   sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
0      3755     2.0           0     0  ...        1610              0   
1      3200     1.5           0     0  ...        1630              0   
2     40705     2.0           0     0  ...        2650              0   
3      6400     1.0           0     0  ...        1240            470   
4      1394     2.0           0     0  ...        1170              0   

   yr_built  yr_renovated  sqft_living15  sqft_lot15  price_per_sq_lot  \
0      1987             0           1300        3823      15106.967566   
1      1912             0           1710        3600      15106.967566   
2      1994             0           2550       42625      15106.967566   
3      1976             0           1600        6400      15106.967566   
4      2001             0           1250        3136      15106.967566   

   delta_renovated  year_sale  month_sale  
0                0       2014           9  
1                0       2014           9  
2                0       2014           6  
3                0       2014          12  
4                0       2014          10  

[5 rows x 22 columns]

2.2 Присоедините к таблице clients данные по домам через метод merge

Это нужно, чтобы понимать, какие дома покупались клиентами

house_id - это индексы датафрейма с домами

In [27]:
df['house_id'] = df.index
df.head()
result = clients.merge(df, on = 'house_id')
result.head()

Empty DataFrame
Columns: [client_id, house_id, id, price, bedrooms, bathrooms, sqft_living, sqft_lot, floors, waterfront, view, condition, grade, sqft_above, sqft_basement, yr_built, yr_renovated, sqft_living15, sqft_lot15, price_per_sq_lot, delta_renovated, year_sale, month_sale]
Index: []

[0 rows x 23 columns]

In [28]:
resalt = clients.merge(df, left_on='house_id', right_on='id')
resalt.head()

client_id  house_id_x          id     price  bedrooms  bathrooms  \
0       1459  8965450190  8965450190  295000.0         3       2.50   
1       4684  6823100225  6823100225  700000.0         4       1.75   
2       3498  5104540330  5104540330  679000.0         4       2.50   
3       3942  2131701075  2131701075  420000.0         3       1.75   
4       4535  1522700060  1522700060  518000.0         4       2.75   

   sqft_living  sqft_lot  floors  waterfront  ...  sqft_basement  yr_built  \
0         1500      3060     2.0           0  ...              0      1994   
1         1870      6000     1.0           0  ...            200      1949   
2         3680      7236     2.0           0  ...              0      2006   
3         1720      5000     1.5           0  ...              0      1932   
4         2520     14021     2.0           0  ...              0      1999   

   yr_renovated  sqft_living15  sqft_lot15  price_per_sq_lot  delta_renovated  \
0             0           2700        7734      15106.967566                0   
1             0           1710        6000      15106.967566                0   
2             0           3310        7180      15106.967566                0   
3          2009           1410        8300      15106.967566               77   
4             0           2330       14007      15106.967566                0   

   year_sale  month_sale  house_id_y  
0       2015           2        7314  
1       2015           4       15505  
2       2015           5        7560  
3       2014          12       13185  
4       2014           6       10395  

[5 rows x 24 columns]

3.1 Найдите среднюю стоимость домов в зависимости от количества спален

 Отсортируйте от меньшей стоимости к большей

In [29]:
df.groupby('bedrooms').agg({'price': 'mean'}).sort_values('price')

price
bedrooms              
1         3.176429e+05
2         4.013727e+05
0         4.095038e+05
3         4.662321e+05
11        5.200000e+05
4         6.354195e+05
33        6.400000e+05
5         7.865998e+05
10        8.193333e+05
6         8.255206e+05
9         8.939998e+05
7         9.511847e+05
8         1.105077e+06

3.2 Найдите минимальную, среднюю и максимальную стоимости домов в зависимости от состояния дома

In [32]:
df.groupby('condition').agg({'price': ['min', 'max', 'mean']})

price                          
                min        max           mean
condition                                    
1           78000.0  1500000.0  334431.666667
2           80000.0  2555000.0  327287.145349
3           75000.0  7062500.0  542012.578148
4           89000.0  7700000.0  521200.390033
5          110000.0  3650000.0  612418.089359

3.3 Постройте таблицу с подсчетом количества домов в данных в зависимости от вида на набережную и оценкой вида

In [33]:
df.pivot_table(index = 'waterfront', columns = 'view', aggfunc='count', values='id', fill_value=0)

view            0    1    2    3    4
waterfront                           
0           19489  331  955  491  184
1               0    1    8   19  135

3.4 Каких домов в зависимости от этажности и количества спален больше?

In [34]:
pd.crosstab(index = df['floors'], columns = df['bedrooms'])

bedrooms  0    1     2     3     4    5    6   7   8   9   10  11  33
floors                                                               
1.0        4  162  1951  5455  2383  605  104   9   5   0   1   0   1
1.5        0   21   182   786   698  185   30   7   1   0   0   0   0
2.0        6   12   497  3118  3682  775  119  19   6   4   2   1   0
2.5        0    1     5    56    58   23   14   2   0   2   0   0   0
3.0        2    3   123   405    61   13    5   1   0   0   0   0   0
3.5        1    0     2     4     0    0    0   0   1   0   0   0   0

3.5 Постройте таблицу с подсчетом медианной стоимости домов в данных в зависимости от состояния дома и оценки дома